### PR TITLE
Azure JSONBlobDataSet feature

### PR DESCRIPTION
**Pull request by kopytjuk**
## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

## Motivation and Context

Since CSV is the only datatype currently supported I wanted to extend currenz `contrib.io.azure` submodule.

## How has this been tested?

Wrote a test suite similar to test_csv.py.

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes
- [x] Assigned myself to the PR
- [x] Added `Type` label to the PR

### CODE REVIEW

File: kedro/contrib/io/azure/json_blob.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+from kedro.io import AbstractDataSet
+
+
+class JSONBlobDataSet(AbstractDataSet):
+    """``JSONBlobDataSet`` loads and saves json(line-delimited) files in Microsoft's Azure
+    blob storage. It uses azure storage SDK to read and write in azure and
+    pandas to handle the json(l) file locally.
+
+    # TODO
```
[the remark this answers is not in the document]
Comment (kopytjuk) [reply kedro/contrib/io/azure/json_blob.py]: Was a todo for writing the usage example - fixed everything now.

File: RELEASE.md [hunk 2]
```diff
@@ -83,6 +83,7 @@ The initial release of Kedro.
 
 ## Thanks for supporting contributions
 
-Jo Stichbury, Aris Valtazanos, Fabian Peters, Guilherme Braccialli, Joel Schwarzmann, Miguel Beltre, Mohammed ElNabawy, Deepyaman Datta, Shubham Agrawal, Oleg Andreyev, Mayur Chougule, William Ashford, Ed Cannon, Nikhilesh Nukala, Sean Bailey, Vikram Tegginamath, Thomas Huijskens, Musa Bilal
+Jo Stichbury, Aris Valtazanos, Fabian Peters, Guilherme Braccialli, Joel Schwarzmann, Miguel Beltre, Mohammed ElNabawy, Deepyaman Datta, Shubham Agrawal, Oleg Andreyev, Mayur Chougule, William Ashford, Ed Cannon, Nikhilesh Nukala, Sean Bailey, Vikram Tegginamath, Thomas Huijskens, Musa Bilal,
```
Comment (idanov) [inline RELEASE.md]:
```suggestion
Jo Stichbury, Aris Valtazanos, Fabian Peters, Guilherme Braccialli, Joel Schwarzmann, Miguel Beltre, Mohammed ElNabawy, Deepyaman Datta, Shubham Agrawal, Oleg Andreyev, Mayur Chougule, William Ashford, Ed Cannon, Nikhilesh Nukala, Sean Bailey, Vikram Tegginamath, Thomas Huijskens, Musa Bilal, Marat Kopytjuk
```

File: RELEASE.md [hunk 3]
```diff
@@ -83,6 +83,7 @@ The initial release of Kedro.
...
+Marat Kopytjuk
```
Comment (idanov) [inline RELEASE.md]:
```suggestion
```

File: kedro/contrib/io/azure/json_blob.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import pandas as pd
+from azure.storage.blob import BlockBlobService
+
+from kedro.io import AbstractDataSet
+
+
+class JSONBlobDataSet(AbstractDataSet):
+    # pylint: disable=too-many-instance-attributes
+    """``JSONBlobDataSet`` loads and saves json(line-delimited) files in Microsoft's Azure
+    blob storage. It uses azure storage SDK to read and write in azure and
```
Comment (idanov) [inline kedro/contrib/io/azure/json_blob.py]:
```suggestion
    blob storage. It uses Azure Storage SDK to read and write in Azure and
```

File: kedro/contrib/io/azure/json_blob.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        save_args: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        """Creates a new instance of ``JSONBlobDataSet`` pointing to a
+        concrete json(l) file on Azure blob storage.
+
+        Args:
+            filepath: path to a azure blob of a json(l) file.
+            container_name: Azure container name.
+            credentials: Credentials (``account_name`` and
+                ``account_key`` or ``sas_token``)to access the azure blob
```
Comment (idanov) [inline kedro/contrib/io/azure/json_blob.py]:
```suggestion
                ``account_key`` or ``sas_token``) to access the Azure blob
```

File: kedro/contrib/io/azure/json_blob.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            encoding: Default utf-8. Defines encoding of json files downloaded as binary streams.
+            blob_to_bytes_args: Any additional arguments to pass to azure's
+                ``get_blob_to_bytes`` method:
+                https://docs.microsoft.com/en-us/python/api/azure.storage.blob.baseblobservice.baseblobservice?view=azure-python#get-blob-to-bytes
+            blob_from_bytes_args: Any additional arguments to pass to azure's
```
Comment (idanov) [inline kedro/contrib/io/azure/json_blob.py]:
```suggestion
            blob_from_bytes_args: Any additional arguments to pass to Azure's
```

File: kedro/template/{{ cookiecutter.repo_name }}/docs/source/conf.py [hunk 7]
```diff
@@ -47,7 +47,7 @@
 #
 import re
 
-from {{ cookiecutter.python_package }} import __version__ as release
```
Comment (idanov) [inline kedro/template/{{ cookiecutter.repo_name }}/docs/source/conf.py]: Could you make sure this PR contains only changes related to the addition of `JSONBlobDataSet`?

File: tests/contrib/io/azure/test_json_blob.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+from unittest.mock import patch
+
+import pandas as pd
+import pytest
+
+from kedro.contrib.io.azure import JSONBlobDataSet
+from kedro.io import DataSetError
+
+TEST_FILE_NAME = "test.csv"
```
Comment (idanov) [inline tests/contrib/io/azure/test_json_blob.py]:
```suggestion
TEST_FILE_NAME = "test.json"
```

File: tests/contrib/io/azure/test_json_blob.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+
+@pytest.fixture()
+def dummy_dataframe():
+    return pd.DataFrame({"col1": [1, 2], "col2": [4, 5], "col3": [5, 6]})
+
+
+@pytest.fixture
+def blob_json_data_set():
+    def make_data_set(load_args=None, save_args=None):
```
Comment (idanov) [inline tests/contrib/io/azure/test_json_blob.py]: If you are simply returning this function, I suppose you can avoid using a fixture here altogether.

File: tests/contrib/io/azure/test_json_blob.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+# pylint: disable=protected-access
+def test_load_extra_params(blob_json_data_set):
+    expected_keys = ["option"]
+    data_set = blob_json_data_set(load_args={"option": "value"})
+    for expected_key in expected_keys:
+        assert expected_key in data_set._load_args.keys()
+
+
+# pylint: disable=protected-access
+def test_save_with_extra_params(blob_json_data_set):
```
Comment (idanov) [inline tests/contrib/io/azure/test_json_blob.py]: This test here isn't testing much either, maybe you can get rid of it as well?

File: tests/contrib/io/azure/test_json_blob.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    blob_from_mock.assert_called_with(
+        container_name=TEST_CONTAINER_NAME,
+        blob_name=TEST_FILE_NAME,
+        blob=dummy_dataframe.to_json().encode("utf-8"),
+        from_extra=42,
+    )
+
+
+# pylint: disable=protected-access
+def test_load_extra_params(blob_json_data_set):
```
Comment (idanov) [inline tests/contrib/io/azure/test_json_blob.py]: This test seems to be testing that python's assignment operator works in the constructor. Could you get rid of it?

File: tests/contrib/io/azure/test_json_blob.py (hunk 9 above)
Comment (kopytjuk) [reply tests/contrib/io/azure/test_json_blob.py]: Found no other easy way to make `make_data_set` a normal helper function. I oriented on the test_csv method from contrib.io.azure.

**Comment by kopytjuk** [comment]
@idanov can you pleae reopen the PR? Thank you!

**Comment by yetudada** [comment]
> @idanov can you pleae reopen the PR? Thank you!

I was wondering what happened. You should be okay to proceed @kopytjuk! 🚀

**Comment by kopytjuk** [comment]
@idanov why does the 3.5 build pipeline fail? I merged the changes from the develop branch but still there is some issue ... https://circleci.com/gh/quantumblacklabs/kedro/551?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Thank you!

**Comment by tolomea** [comment]
@kopytjuk The failure is not related to your changes, it's an intermittent problem we are having across the branches. I suspect it's related to memory use on the Circle worker but I'm not sure yet.

File: kedro/contrib/io/azure/json_blob.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        self._load_args = load_args if load_args else {}
+        self._filepath = filepath
+        self._encoding = encoding
+        self._container_name = container_name
+        self._credentials = credentials if credentials else {}
+        self._blob_to_bytes_args = blob_to_bytes_args if blob_to_bytes_args else {}
+        self._blob_from_bytes_args = blob_from_bytes_args if blob_from_bytes_args else {}
+
+    def _load(self) -> pd.DataFrame:
+
```
Comment (921kiyo) [inline kedro/contrib/io/azure/json_blob.py]:
```suggestion
```

File: kedro/contrib/io/azure/json_blob.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        blob_service = BlockBlobService(**self._credentials)
+
```
Comment (921kiyo) [inline kedro/contrib/io/azure/json_blob.py]:
```suggestion
```

File: kedro/contrib/io/azure/json_blob.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        blob = blob_service.get_blob_to_bytes(
+            container_name=self._container_name, blob_name=self._filepath,
+            **self._blob_to_bytes_args)
+        bytes_stream = io.BytesIO(blob.content)
+
```
Comment (921kiyo) [inline kedro/contrib/io/azure/json_blob.py]:
```suggestion
```

File: kedro/contrib/io/azure/json_blob.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        return pd.read_json(bytes_stream, encoding=self._encoding, **self._load_args)
+
+    def _save(self, data: pd.DataFrame) -> None:
+        blob_service = BlockBlobService(**self._credentials)
+
```
Comment (921kiyo) [inline kedro/contrib/io/azure/json_blob.py]:
```suggestion
```

**Pull request merged**